pczt/src/orchard.rs

```diff
@@ -0,0 +1,535 @@
+    #[serde_as(as = "Option<[_; 43]>")]
+    #[getset(get = "pub")]
+    pub(crate) recipient: Option<[u8; 43]>,
+
+    /// The value of the output.
+    ///
+    /// This may be used by Signers to verify that the value matches `cv`, and to confirm
+    /// the values and change involved in the transaction.
+    ///
+    /// This exposes the value to all participants. For Signers who don't need this
+    /// information, we can drop the values and compress the rcvs into the bsk global.
+    #[getset(get = "pub")]
+    pub(crate) value: Option<u64>,
+
+    /// The seed randomness for the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover, instead of disclosing `shared_secret` to them.
+    #[getset(get = "pub")]
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The `ock` value used to encrypt `out_ciphertext`.
+    ///
+    /// This enables Signers to verify that `out_ciphertext` is correctly encrypted.
+    ///
+    /// This may be `None` if the Constructor added the output using an OVK policy of
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being created.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut actions,
+            flags,
+            value_sum,
+            anchor,
+            zkproof,
+            bsk,
+        } = other;
+
+        if self.flags != flags {
+            return None;
+        }
+
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of actions, and the value sums must match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
+                if self.actions.len() != actions.len() || self.value_sum != value_sum =>
+            {
+                return None
+            }
+            // IO Finalizer has run, and neither bundle has excess spends or outputs.
+            (Some(_), _) | (_, Some(_)) => (),
+            // IO Finalizer has not run on either bundle. If the other bundle has more
+            // spends or outputs than us, move them over; these cannot conflict by
+            // construction.
+            (None, None) => {
+                if actions.len() > self.actions.len() {
+                    self.actions.extend(actions.drain(self.actions.len()..));
```

Comment:
This does not check whether we have any signatures. If we do then modifying `self.actions` would invalidate them.

Reply:
Checking the `tx_modifiable` flags, and relying on the Signer to update them correctly, should be sufficient here. (If they don't, the only effect is an identifiable DoS because the final tx extraction will fail at signature validation.)